pkg/compact/compact.go

```diff
@@ -1318,44 +1324,89 @@ func (c *BucketCompactor) Compact(ctx context.Context) (rerr error) {
 
 		level.Info(c.logger).Log("msg", "start of compactions")
 
-		// Send all groups found during this pass to the compaction workers.
-		var groupErrs errutil.MultiError
-	groupLoop:
-		for _, g := range groups {
-			// Ignore groups with only one block because there is nothing to compact.
-			if len(g.IDs()) == 1 {
-				continue
-			}
+		tasks, err := c.planTasks(ctx, groups)
+		if err != nil {
+			return err
+		}
+
+		// Send all tasks planned in this pass to the compaction workers.
+		var taskErrs errutil.MultiError
+	tasksLoop:
+		for _, task := range tasks {
 			select {
-			case groupErr := <-errChan:
-				groupErrs.Add(groupErr)
-				break groupLoop
-			case groupChan <- g:
+			case taskErr := <-errChan:
+				taskErrs.Add(taskErr)
+				break tasksLoop
+			case taskChan <- task:
 			}
 		}
-		close(groupChan)
+		close(taskChan)
 		wg.Wait()
 
 		// Collect any other error reported by the workers, or any error reported
 		// while we were waiting for the last batch of groups to run the compaction.
 		close(errChan)
 		for groupErr := range errChan {
-			groupErrs.Add(groupErr)
+			taskErrs.Add(groupErr)
 		}
 
 		workCtxCancel()
-		if len(groupErrs) > 0 {
-			return groupErrs.Err()
+		if len(taskErrs) > 0 {
+			return taskErrs.Err()
 		}
 
-		if finishedAllGroups {
+		if finishedAllTasks {
 			break
 		}
 	}
 	level.Info(c.logger).Log("msg", "compaction iterations done")
 	return nil
 }
 
+func (c *BucketCompactor) planTasks(ctx context.Context, groups []*Group) ([]GroupCompactionTask, error) {
+	// Plan tasks from all groups
+	allTasks := make([][]GroupCompactionTask, 0, len(groups))
+	numTasks := 0
+	for _, g := range groups {
+		// Ignore groups with only one block because there is nothing to compact.
+		if len(g.IDs()) == 1 {
+			continue
+		}
+
+		groupTasks, err := g.PlanCompactionTasks(ctx, c.compactDir, c.planner)
+		if err != nil {
+			return nil, errors.Wrapf(err, "get compaction group tasks: %s", g.Key())
+		}
+		allTasks = append(allTasks, groupTasks)
+		numTasks += len(groupTasks)
+	}
+
+	tasksLimit := c.concurrency
+	// Make sure we plan at least one task from each group.
+	if tasksLimit < len(allTasks) {
+		tasksLimit = len(groups)
```

Comment:
Is it `tasksLimit = len(groups)` or `tasksLimit = len(allTasks)`?

Reply:
You're right, this should be `len(allTasks)` because we can have empty compaction groups.